File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int result = checkHeight(root);

        if(result == -10000) {
            return false;
        }

        return true;
    }

    int checkHeight(TreeNode* root) {
        if(root == nullptr) {
            return -1;
        }

        int left = checkHeight(root->left);

        if(left == -10000) return -10000;

        int right = checkHeight(root->right);

        if(right == -10000) return -10000;       

        int diff = right - left;

        if(!(diff >= -1 && diff <= 1)) {
            return -10000;
        }

        return max(left, right) + 1;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public int goodNodes(TreeNode root) {
        return goodNodesHelper(root, root.val);
    }

    public int goodNodesHelper(TreeNode root,  int max) {
        if(root == null) {
            return 0;
        }

        int c = 0;

        if(root.val >= max) {
            c++;
        }

        int left = goodNodesHelper(root.left, Math.max(root.val, max));
        int right = goodNodesHelper(root.right, Math.max(root.val, max));

        return c + left + right;        
    }
}